Fix getValueFromArraybuffer calls to use proper type

Achie72 opened the pull request:
Fixes: #225

Signed-off-by: Bela Toth tbela@inf.u-szeged.hu

rerobika commented on src/runtime/GlobalObjectBuiltinTypedArray.cpp:
```diff
@@ -456,7 +456,38 @@ Value builtinTypedArrayConstructor(ExecutionState& state, Value thisValue, size_
                     // Let value be GetValueFromBuffer(srcData, srcByteIndex, srcType).
-                    Value value = srcData->getValueFromBuffer<typename TypeAdaptor::Type>(state, srcByteIndex);
+                    Value value;
+                    switch (val.asObject()->asArrayBufferView()->typedArrayType()) {
```
Could you please separate it to an utility function for further reusability?

clover2123 commented:
Would you please add a internal TC related to #225?

Achie72 commented:
@rerobika I think the separation is done, and should work properly.

@clover2123 I added a TC for the original problem, which was caused by improper reading and copying of data inside the buffer objects. I added a few cases testing that as well (copying proper value, handling overflow and signed-unsigned conversions). I hope it's not a problem.

rerobika commented on src/runtime/ArrayBufferObject.h:
```diff
@@ -98,6 +110,43 @@ class ArrayBufferObject : public Object {
+        Value value;
+        switch (type) {
+        case TypedArrayType::Int8:
+            value = getValueFromBuffer<int8_t>(state, byteindex, true);
```
I think you can directly return in all cases.

Achie72 replied:
That's a nice catch, that flew past by me. I updated the PR with this change.

rerobika replied:
I think this code path has not been updated.

rerobika commented on src/runtime/ArrayBufferObject.h:
```diff
@@ -98,6 +110,34 @@ class ArrayBufferObject : public Object {
 
+    Value getTypedValueFromBuffer(ExecutionState& state, unsigned byteindex, TypedArrayType type)
+    {
+        Value value;
```
It seems to me that this variable is unused.

rerobika commented:
@Achie72 please rebase this PR.
